Rebuild related changes based on metadata

mynktl opened the pull request:
- API to get modified data from metadata
    - Handling of rebuild IO from other replica according to metadata
    - Removing old rebuilding code (txg base rebuilding)

Commit:
- API to get modified data from metadata
- Handling of rebuild IO from other replica according to metadata
- Removing code based on txg number

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

satbirchhikara commented on lib/libzpool/uzfs_io.c:
```diff
@@ -99,15 +99,12 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
-			    len, (void **)&chunk_io);
-			if (!count)
-				goto exit_with_error;
+	if ((zv->zv_status & ZVOL_REBUILDING_IN_PROGRESS) &&
```
zv_status is a variable to maintain state not flags. For improving readability better to use  "=="/"logical equal" to operator.

mynktl replied:
ok

satbirchhikara reviewed (commented):
Hi Mayank,

I have not looked in to changes yet. Just want to check if we missing data API will work on given range ?

Idea is that dowgrade replica may pass <io_seq, start offset of scan, len/end offset of scan> to healthy replica, 
asking healthy replica to just scan data between <start offset of scan> and <len/end offset of scan>.

This is required where we want to rebuild a downgraded replica from multiple healthy replicas for different ranges in LUN. 

One usecase where it can make sense is assume, a new replica has been added to cluster in that case, instead of rebuilding from one healthy replica, we can partition LUN in smaller ranges/chunks and can rebuild from different healthy replicas.

I am not sure if but if fresh replica added, we may need an optimized way of copying data from healthy replicas to this newly added replica by avoiding some of the checks.

Commit:
- Changes in uzfs_get_io_diff to get metadata based on offset and length
  Snapshot created with provided metadata will be removed if (offset + length) == (end of metadata)
- Replaced dsl_pool_hold with dmu_object_own to hold snapshot in uzfs_get_io_diff
  Since, it was causing deadlock
- Functionality test for partial read from uzfs_get_io_diff

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

gila reviewed (approved):
Looks good as usual

jkryl commented on include/sys/uzfs_zvol.h:
```diff
@@ -53,9 +53,7 @@ typedef struct metaobj_blk_offset {
-	kmutex_t io_tree_mtx;
 	uint64_t rebuild_bytes;
-	avl_tree_t *incoming_io_tree;   /* incoming io tree */
 } zvol_rebuild_data_t;
```
should this structure be removed? (there is just one number inside and stale comment)

mynktl replied:
Thanks for pointing out this.
```zvol_rebuild_data_t``` was meant to hold all rebuild related variables. So better to keep as it is.

As of now, there are two variables we are using for rebuild purpose ```rebuild_bytes``` and ```zv_rebuild_status```. I will move ```zv_rebuild_status``` to ```zvol_rebuild_data_t``` structure.

I will also remove stale comment.

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,296 @@
+				last_lun_offset = lun_offset;
+			}
+			diff_count++;
+		} else if (!ret) {
```
this should never happen in our code path

mynktl replied:
This is not required if the single write has single io number.
I will merge this case with on_disk_io_number > rebuild io number.

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,296 @@
+		    metadatasize;
+		ret = compare_blk_metadata(rd_metadata, metadata);
+		if (ret == -1) {
+			// old io number < new io number
```
on_disk io number < incoming IO number

mynktl replied:
Ok. that makes sense. I will update comment.

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,296 @@
+		ret = compare_blk_metadata(rd_metadata, metadata);
+		if (ret == -1) {
+			// old io number < new io number
+			if (diff_count == 0) {
```
no need of curly braces..

mynktl replied:
ok

vishnuitta commented on include/uzfs_rebuilding.h:
```diff
@@ -0,0 +1,51 @@
+ * than w_metadata then that part of IO's will be discarded else it will be
+ * added to list.
+ */
+int uzfs_search_nonoverlapping_io(zvol_state_t *zv, uint64_t offset,
```
uzfs_get_condensed_old_ondisk_blks..?

mynktl replied:
ok

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -310,3 +308,52 @@ uzfs_zvol_get_rebuild_status(zvol_state_t *zv)
 }
+
+/*
+ * This assumes snapshot, so, no need of range lock
```
this doesn't seems to be the case all the time

mynktl replied:
yes. Actually, range lock for metadata is not needed since all the operation on metadata will go through ```uzfs_read_data``` and ```uzfs_write_data```.
In uzfs_get_io_diff, we are calling uzfs_read_metadata but that is on snapshot so no need of range lock.
I will remove range lock for metadata.

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -310,3 +308,52 @@ uzfs_zvol_get_rebuild_status(zvol_state_t *zv)
+	uint64_t blocksize = zv->zv_volmetablocksize;
+	uint64_t len_in_first_aligned_block, bytes, read = 0;
+	uint64_t end = offset + len;
+	uint64_t metaobjectsize = P2ALIGN(zv->zv_volsize,
```
this doesn't seems to be correct.. we need to care of metadatasize also..

mynktl replied:
yes. I will update this.

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -310,3 +308,52 @@ uzfs_zvol_get_rebuild_status(zvol_state_t *zv)
+			bytes = metaobjectsize - offset;
+
+		ret = dmu_read(zv->zv_objset, ZVOL_META_OBJ, offset, bytes,
+		    buf + read, 0);
```
should we also have a check that we are not writing more than allocated space?

vishnuitta replied:
I think - not required.. this should be good

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,296 @@
+
+int
+uzfs_get_io_diff(zvol_state_t *zv, blk_metadata_t *low,
+    uzfs_get_io_diff_cb_t *func, off_t zvol_offset, size_t zvol_len,
```
make zvol_offset, zvol_len as lun_offset, lun_len

mynktl replied:
ok.

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,296 @@
+		}
+
+		strfree(snap_name);
+		ret = dmu_objset_own(dataset, DMU_OST_ANY, B_TRUE, snap_zv,
```
how about using uzfs_open_dataset_init, which creates and initializes zv..?

mynktl replied:
yes. we can use it. I will try that and update changes.

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,296 @@
+	if (end > metaobjectsize)
+		end = metaobjectsize;
+
+	ret = get_metadata_snapshot_info(zv, low, &snap_zv);
```
get_snapshot_zv

mynktl replied:
ok. I will make this function generic as ```get_snapshot_zv(zv, snap_name, snap_zv)```

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,296 @@
+					last_index = i;
+				}
+				diff_count++;
+				if (last_md != NULL &&
```
check on last_md is not required. add some comments for this logic.

mynktl replied:
ok

vishnuitta commented on include/uzfs_io.h:
```diff
@@ -56,4 +56,9 @@ extern int uzfs_zvol_get_rebuild_status(void *zv);
+/*
+ * API to read metadata
+ */
+extern int uzfs_read_metadata(void *zv, char *buf, uint64_t offset,
```
make zv as zvol_state_t

vishnuitta replied:
as these APIs are being used in production also, having strict check on arguments will be good

mynktl replied:
ok

vishnuitta commented on include/sys/uzfs_zvol.h:
```diff
@@ -149,11 +147,12 @@ typedef struct uzfs_zvol_blk_phy {
-typedef int (uzfs_txg_diff_traverse_cb_t)(off_t offset, size_t len,
-    uint64_t blkid, void *arg);
+typedef int (uzfs_get_io_diff_cb_t)(off_t offset, size_t len,
+    blk_metadata_t *metadata, objset_t *obj, void *arg);
```
how about passing zvol_state_t and using uzfs_read_data rather than dmu_read in callback implementation?

mynktl replied:
ok

vishnuitta reviewed (commented):
overall, looks good.
metaobjectsize issue in uzfs_read_metadata should have caught in unit testing code. probably, higher active data size may be required.

Commit:
- resolving conflicts with zfs-0.7-release
- code changes according to review comments

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

mynktl committed:
Merge branch 'zfs-0.7-release' into cstor64

vishnuitta commented on lib/libzpool/uzfs_io.c:
```diff
@@ -48,14 +49,9 @@
 
 #define	WRITE_METADATA(zv, metablk, mdata, tx)			\
 	do {							\
-		rl_t *mrl;					\
```
this will be good if offset and len are multiple of zv_metavolblocksize. And, uzfs also recevies IOs in same way from tgt.
So, please add VERIFY in read and write APIs for offset & len to be multiple of zv_metavolblocksize

mynktl replied:
ok

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,293 @@
+	dataset = kmem_asprintf("%s@%s", strchr(zv->zv_name, '/') + 1,
+	    snap_name);
+
+	ret = uzfs_open_dataset(zv->zv_spa, dataset, snap_zv);
```
uzfs_open_dataset calls uzfs_open_dataset_init.. please change uzfs_open_dataset_init as uzfs_own_dataset

mynktl replied:
ok

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,293 @@
+				// same as last one.
+				diff_count++;
+
+				if (compare_blk_metadata((blk_metadata_t *)
```
diff_count++ should go into 'else' case, and, last_lun_offset/last_md/last_index need to be set in 'if' after EXECUTE_... macro

mynktl replied:
ok

vishnuitta commented on lib/libzpool/uzfs_rebuilding.c:
```diff
@@ -0,0 +1,293 @@
+			lun_offset += zv->zv_metavolblocksize;
+		}
+
+		if (diff_count) {
```
&&!ret

mynktl replied:
it is not needed here

Commit:
- fix in uzfs_get_io_diff API
- changes according to review comments
- uzfs_open_dataset_init changed to uzfs_own_dataset for better
  readability
Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
Merge branch 'cstor64' of github.com:mynktl/zfs into cstor64

Commit:
- reducing number of tests for rebuilding
- setting metavolblocksize to 512 bytes when creating volume through CLI

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta reviewed (approved):
looks good
